data_preparation/predictor_data_prep.py

```diff
@@ -0,0 +1,1137 @@
+        mask,
+        transform=out_transform,
+        invert=mask_invert,
+        out_shape=out_shape,
+        all_touched=mask_all_touched,
+    )
+    # write the mask output
+    if out_mask_path:
+        with rasterio.open(
+            out_mask_path,
+            "w",
+            driver="Gtiff",
+            height=input_geom_mask.shape[0],
+            width=input_geom_mask.shape[1],
+            count=1,
+            dtype=np.uint8,
+            crs=dataset.crs,
+            transform=out_transform,
+        ) as tmp_dataset:
+            tmp_dataset.write(input_geom_mask.astype(np.uint8), 1)
+
+    # mask data arrays
+    source = np.ma.array(source, mask=input_geom_mask)
+
+    return source
+
+
+"""
+@author: Joint Research Centre - D5 Food Security - ASAP
+"""
+
+
+def round_window(window):
+    """
+    Outputs a copy of the window rounded to the nearest whole pixel.
+    Rounds the absolute size of the window extent which makes sure that all pixel cneters covered by the input window
+    are included.
+    :param window: Input window
+    :return: rasterio.windows.Window
+    """
+    _col_off = round(window.col_off)
+    _row_off = round(window.row_off)
+    _width = round(window.col_off + window.width) - _col_off
+    _height = round(window.row_off + window.height) - _row_off
+    return rasterio.windows.Window(_col_off, _row_off, _width, _height)
+
+
+"""
+@author: Joint Research Centre - D5 Food Security - ASAP
+"""
+
+
+def get_common_bounds_and_shape(geom, ds_list):
+    """
+    Returns a unified read window, bounds and resolution, for a given geometry on all raster datasets in a list.
+    Used when combining datasets with heterogeneous resolution to read data with the extent (bounds) of the lowest
+    resolution dataset and to use the resolution (shape) of the highest resolution dataset.
+    Useful when trying to get unified arrays to perform calculation from heterogeneous datasets.
+    The read window and resolution is defined with a following process.
+        1. we find the lowest and highest resolution datasets by comparing their geometry window sizes (all pixels
+           touched by geom).
+        2. we get the geometry window bounds of the lowest resolution dataset and align them to the highest
+           resolution dataset.
+            - lowest resolution dataset window should be the maximum extent, extents covering all other dataset
+              windows.
+            - We align the bounds to the highest resolution dataset by defining the bounds window on the hres
+              dataset and rounding them to the closest whole pixel
+        3. we define unified read window and out_shape
+            - We use the aligned bounds to define a read window for all datasets
+            - We use the resolution of the hres rounded bounds window shape as the read out_shape
+    :param geom: GeoJSON-like feature (implements __geo_interface__) – feature collection, or geometry.
```

Comment:
@umdsgy can you please update the input parameter description to match the change from the original code. "mask" parameter changed form geometry -> iterable over geometries